src/app-layout/utils/use-resize.tsx

```diff
@@ -0,0 +1,134 @@
+
+import splitPanelStyles from '../../split-panel/styles.css.js';
+import testutilStyles from '../test-classes/styles.css.js';
+import styles from '../visual-refresh/styles.css.js';
+import { DrawerFocusControlRefs } from './use-drawer-focus-control';
+
+export interface SizeControlProps {
+  position: 'side';
+  panelRef?: React.RefObject<HTMLDivElement>;
+  handleRef?: React.RefObject<HTMLDivElement>;
+  setSidePanelWidth: (width: number) => void;
+  setBottomPanelHeight: (height: number) => void;
+}
+
+export interface DrawerResizeProps {
+  activeDrawerId: string | null;
+  drawers: DrawersProps;
+  drawersRefs: DrawerFocusControlRefs;
+  isToolsOpen: boolean;
+  drawersMaxWidth: number;
+}
+
+function useResize(drawerRefObject: React.RefObject<HTMLDivElement>, drawerResizeProps: DrawerResizeProps) {
+  const { activeDrawerId, drawers, drawersRefs, isToolsOpen, drawersMaxWidth } = drawerResizeProps;
+
+  const activeDrawer = drawers?.items.find((item: any) => item.id === activeDrawerId) ?? null;
+  const drawerItems = useMemo(() => drawers?.items || [], [drawers?.items]);
+  const toolsWidth = 290;
+  const MIN_WIDTH = activeDrawer?.defaultSize && activeDrawer.defaultSize < 290 ? activeDrawer?.defaultSize : 290;
+  const [relativeSize, setRelativeSize] = useState(0);
+  const getDrawerItemSizes = useCallback(() => {
+    const sizes: { [id: string]: number } = {};
+    if (!drawerItems) {
+      return {};
+    }
+
+    for (const item of drawerItems) {
+      if (item.defaultSize) {
+        if (item.defaultSize > drawersMaxWidth) {
+          sizes[item.id] = toolsWidth;
+        } else {
+          sizes[item.id] = item.defaultSize || toolsWidth;
+        }
+      }
+    }
+    return sizes;
+  }, [drawerItems, toolsWidth, drawersMaxWidth]);
+  const [drawerItemSizes, setDrawerItemSizes] = useState(() => getDrawerItemSizes());
+
+  const drawerSize =
+    !activeDrawerId && !isToolsOpen
+      ? 0
+      : activeDrawerId && drawerItemSizes[activeDrawerId]
+      ? drawerItemSizes[activeDrawerId]
+      : toolsWidth;
+
+  useEffect(() => {
+    // Ensure we only set new drawer items by performing a shallow merge
+    // of the latest drawer item sizes, and previous drawer item sizes.
+    setDrawerItemSizes(() => getDrawerItemSizes());
+  }, [drawersMaxWidth, activeDrawerId, getDrawerItemSizes]);
+
+  useEffect(() => {
+    // effects are called inside out in the components tree
+    // wait one frame to allow app-layout to complete its calculations
+    const handle = requestAnimationFrame(() => {
+      const maxSize = drawersMaxWidth;
+      setRelativeSize(((drawerSize - MIN_WIDTH) / (maxSize - MIN_WIDTH)) * 100);
+    });
+    return () => cancelAnimationFrame(handle);
+  }, [drawerSize, drawersMaxWidth, MIN_WIDTH]);
+
+  const drawerResize = (resizeDetail: { size: number; id: string }) => {
+    const drawerItem = drawers.items.find(({ id }: any) => id === resizeDetail.id);
+    if (drawerItem?.onResize) {
```

Comment:
[`fireNonCancelableEvent` checks internally if the handler is undefined](https://github.com/cloudscape-design/components/blob/d5aa98e030232223a427f2f927e4ac4b91d6c162/src/internal/events/index.ts#L55-L57), so this condition here can be skipped